sentry/src/routes/campaign.rs

```diff
@@ -25,32 +68,197 @@ pub async fn create_campaign<A: Adapter>(
+
+    // TODO: AIP#61: Update when changes to Spendable are ready
+    let latest_spendable = fetch_spendable(app.pool.clone(), &campaign.creator, &campaign.channel.id()).await?;
+    let remaining_for_channel = get_total_remaining_for_channel(&accounting_spent, &latest_spendable).ok_or(ResponseError::BadRequest("couldn't get total remaining for channel".to_string()))?;
 
-    // match insert_campaign(&app.pool, &campaign).await {
-    //     Err(error) => {
-    //         error!(&app.logger, "{}", &error; "module" => "create_channel");
-
-    //         match error {
-    //             PoolError::Backend(error) if error.code() == Some(&SqlState::UNIQUE_VIOLATION) => {
-    //                 Err(ResponseError::Conflict(
-    //                     "channel already exists".to_string(),
-    //                 ))
-    //             }
-    //             _ => Err(error_response),
-    //         }
-    //     }
-    //     Ok(false) => Err(error_response),
-    //     _ => Ok(()),
-    // }?;
-
-    let create_response = SuccessResponse { success: true };
+    if campaign.budget > remaining_for_channel {
+        return Err(ResponseError::BadRequest("Not Enough budget for campaign".to_string()));
+    }
+
+    // If the channel is being created, the amount spent is 0, therefore remaining = budget
+    increase_remaining_for_campaign(&app.redis.clone(), campaign.id, campaign.budget).await.map_err(|_| ResponseError::BadRequest("Couldn't update remaining while creating campaign".to_string()))?;
+
+    // insert Campaign
+    match insert_campaign(&app.pool, &campaign).await {
+        Err(error) => {
+            error!(&app.logger, "{}", &error; "module" => "create_campaign");
+            match error {
+                PoolError::Backend(error) if error.code() == Some(&SqlState::UNIQUE_VIOLATION) => {
+                    Err(ResponseError::Conflict(
+                        "Campaign already exists".to_string(),
+                    ))
+                }
+                _ => Err(error_response),
+            }
+        }
+        Ok(false) => Err(ResponseError::BadRequest("Encountered error while creating Campaign; please try again".to_string())),
+        _ => Ok(()),
+    }?;
 
     Ok(success_response(serde_json::to_string(&campaign)?))
 }
 
+
+// tested
+fn get_total_remaining_for_channel(accounting_spent: &UnifiedNum, latest_spendable: &Spendable) -> Option<UnifiedNum> {
+    let total_deposited = latest_spendable.deposit.total;
+
+    let total_remaining = total_deposited.checked_sub(&accounting_spent);
+    total_remaining
+}
+
+pub mod update_campaign {
+    use super::*;
+    use lazy_static::lazy_static;
+
+    lazy_static!{
+        pub static ref CAMPAIGN_REMAINING_KEY: &'static str = "campaignRemaining";
+    }
+
+    pub async fn increase_remaining_for_campaign(redis: &MultiplexedConnection, id: CampaignId, amount: UnifiedNum) -> Result<bool, CampaignError> {
```

Comment:
I'm hesitant if we should use `INCRBY` for `Campaign` creation. It makes more sense to fail if the campaign Id is already in Redis.

We can use the `SET` with the option `NX -- Only set the key if it does not already exist` and return an error if there is value already for this `CampaignId`.

https://redis.io/commands/set

How does it sound to you @simzzz ?

Reply:
Yes, that would be a better solution.